use-timescale/compression/compression-design.md

```diff
@@ -0,0 +1,176 @@
+column. For example, if you wanted to store temperatures that changed over time
+for two different devices, and you used delta encoding on the time column, you
+might end up with a dataset that looks like this:
+
+|Timestamp|Device ID|Status Code|Temperature|
+|-|-|-|-|
+|12:00:01|A|0|70.11|
+|1 second|A|0|70.12|
+|1 second|A|0|70.14|
+|12:00:01|B|0|69.70|
+|1 second|B|0|69.69|
+|1 second|B|4|69.70|
+
+Compressing in this way is fairly effective, because timestamps are stored at
+microsecond precision and are represented as a 64-bit integer, so this allows
+the timestamp column to be stored as 64 bits for the time and the increments
+using 20 bits for a total of 26 bytes rather than 48 bytes, so the compression
+rate for this data is 46%.
+
+TimescaleDB uses a more advanced compression algorithm for timestamps called
+[delta-delta encoding][delta-delta] which gives even better compression ratio for
+timestamps, but delta encoding is used in this section for illustrative purposes.
+
+However, looking closely at the example, you can see the data is ordered by
+device ID, not by time. This can happen if you insert data into the hypertable
+in batches, but it can also create a problem for the compression algorithms. If
+the data is ordered by timestamp instead of by device ID, the compressed data
+looks like this instead:
+
+|Timestamp|Device ID|Status Code|Temperature|
+|-|-|-|-|
+|12:00:01|A|0|70.11|
+|0 seconds|B|0|69.70|
+|1 second|A|0|70.12|
+|0 seconds|B|0|69.69|
+|1 second|A|0|70.14|
+|0 seconds|B|4|69.70|
+
+This compresses the timestamps to 64 bits plus 2 * 20 bits, which is 73%
+compression, and much higher than the first example, for exactly the same data.
+
+By default, TimescaleDB orders data by time, but you can also add additional
+columns for ordering.
+
+## Segmenting compressed data
+
+When you have compressed your data in the way described in the previous section,
+the database has no way of knowing which rows need to be fetched and
+decompressed to resolve a query. For example, the database can't easily
+determine which rows contain data from the past day, as the timestamp itself is
+in a compressed column. You don't want to have to decompress all the data in a
+chunk, or even an entire hypertable, to determine which rows are required.
+TimescaleDB automatically includes more information in each batch to improve
+query performance.
+
+For order-by columns, TimescaleDB automatically adds additional data that stores
+the minimum and maximum value of that column. This way, the query planner can
+look at the range of timestamps in the compressed column, without having to
+decompress a batch, and determine whether the data in the batch could possibly
+match the query.
+
+However, using the previous example, if you wanted to query the data based on
+device ID, you would have to decompress the device column to decide if that
+batch should be part of the query. When you compress your hypertable, you can
```

Comment:
It is not entirely clear why the entire column have to be decompressed.

One reason is because every compressed row is likely to contain every device, which means that having a max and a min does not help in deciding if the row need to be decompressed: the test indicates that the column always has to be decompressed.

The other reason is because we cannot have an index on data in a compressed column.

Reply:
My interpretation was the former: that you don't know if the device is referenced in the column without decompressing it first.

Reply:
Sorry, I was unclear. Both reasons are valid, but let me elaborate a little.

Typically, device IDs are taken from a separate table containing a serial ID. This means that device IDs typically are integers, not strings. With this in mind, we can compress the device ID column as well, which gives us something like this:

| `device`              | `_ts_meta_min_1` | `_ts_meta_max_1` |
|-----------------------|------------------|------------------|
| [0, 1, 2, 1, 2, 3, 0] | 0                | 3                |
| [0, 0, 3, 0, 0, 2, 0] | 0                | 3                |

However, as you can see, the range does not allow us to filter out the rows that does not contain 1 since the entire range of device IDs are covered in each row, so we have to scan all rows and decompress the column to see if there is a match.

However, if we had an index on the device ID, we would be able to select all the compressed rows for that device, so by segmenting the data by the device ID, we can easily select the rows that have matching IDs using an index.